File: api/models/item_lines.py
import json

from models.base import Base
from providers import data_provider

ITEM_LINES = []


class ItemLines(Base):
    def __init__(self, root_path, is_debug=False):
        self.data_path = root_path + "item_lines.json"
        self.load(is_debug)

    def get_item_lines(self):
        return self.data

    def get_item_line(self, item_line_id):
        for x in self.data:
            if x["id"] == item_line_id:
                return x
        return None

    def add_item_line(self, item_line):
        for x in self.data:
            if x["id"] == item_line["id"]:
                return False
        item_line["created_at"] = self.get_timestamp()
        item_line["updated_at"] = self.get_timestamp()
        self.data.append(item_line)
        return True

    def update_item_line(self, item_line_id, item_line):
        if "id" in item_line:
            if item_line_id != item_line["id"]:
                return False
        item_line["updated_at"] = self.get_timestamp()
        for i in range(len(self.data)):
            if self.data[i]["id"] == item_line_id:
                self.data[i] = item_line
                return True

    def remove_item_line(self, item_line_id):
        item_line = self.get_item_line(item_line_id)
        if item_line is None: return False
        items = data_provider.fetch_item_pool().get_items_for_item_line(item_line_id)
        if len(items) != 0: return False
        for x in self.data:
            if x["id"] == item_line_id:
                self.data.remove(x)
                return True

    def load(self, is_debug):
        if is_debug:
            self.data = ITEM_LINES
        else:
            f = open(self.data_path, "r")
            self.data = json.load(f)
            f.close()

    def save(self):
        f = open(self.data_path, "w")
        json.dump(self.data, f)
        f.close()

File: C#api/models/item_lines.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;
using Newtonsoft.Json;
using Providers;
namespace Models;


public class ItemLine
{
    public int? Id { get; set; }
    public string Name { get; set; }
    public string Description { get; set; }
    public string? Created_At { get; set; }
    public string? Updated_At { get; set; }
}

public class ItemLines : Base
{
    private readonly string _dataPath;
    private List<ItemLine> _data;

    public ItemLines(string rootPath, bool isDebug = false)
    {
        _dataPath = Path.Combine(rootPath, "item_lines.json");
        Load(isDebug);
    }

    public List<ItemLine> GetItemLines()
    {
        return _data;
    }

    public ItemLine GetItemLine(int itemlineId)
    {
        return _data.Find(x => x.Id == itemlineId);
    }

    public bool AddItemline(ItemLine itemline)
    {
        itemline.Id = _data.Count > 0 ? _data.Max(il => il.Id) + 1 : 1;
        if (itemline.Created_At == null) itemline.Created_At = GetTimestamp();
        if (itemline.Updated_At == null) itemline.Updated_At = GetTimestamp();
        _data.Add(itemline);
        return true;
    }

    public bool UpdateItemline(int itemlineId, ItemLine itemline)
    {
        itemline.Updated_At = GetTimestamp();
        var index = _data.FindIndex(x => x.Id == itemlineId);
        if (index >= 0)
        {
            itemline.Id = _data[index].Id;
            itemline.Created_At = _data[index].Created_At;
            _data[index] = itemline;
            return true;
        }

        return false;
    }

    public bool ReplaceItemLine(int itemLineId , ItemLine newItemLine)
    {
        var index = _data.FindIndex(existingItemLine => existingItemLine.Id == itemLineId);
        var existingItemLine = _data.FirstOrDefault(existingItemLine => existingItemLine.Id == itemLineId);

        if (index < 0)
        {

            return false;

        }

        if (!string.IsNullOrEmpty(newItemLine.Name)) existingItemLine.Name = newItemLine.Name;
        if (!string.IsNullOrEmpty(newItemLine.Description)) existingItemLine.Description = newItemLine.Description;
        existingItemLine.Updated_At = GetTimestamp();

        return true;
    }

    public bool RemoveItemline(int itemlineId, bool force = false)
    {
        var itemline = GetItemLine(itemlineId);
        if (itemline == null) return false;
        if (force) return _data.Remove(itemline);
        var items = DataProvider.fetch_item_pool().GetItemsForItemLine(itemlineId);
        if (items.Count() != 0) return false;

        return _data.Remove(itemline);
    }

    private void Load(bool isDebug)
    {
        if (isDebug)
        {
            _data = new List<ItemLine>();
        }
        else
        {
            using (var reader = new StreamReader(_dataPath))
            {
                var json = reader.ReadToEnd();
                _data = JsonConvert.DeserializeObject<List<ItemLine>>(json);
            }
        }
    }

    public void Save()
    {
        using (var writer = new StreamWriter(_dataPath))
        {
            var json = JsonConvert.SerializeObject(_data, Formatting.Indented);
            writer.Write(json);
        }
    }
}
